cmd/goal/application.go

```diff
@@ -995,3 +1044,141 @@ var infoAppCmd = &cobra.Command{
+
+		// Construct schemas from args
+		localSchema := basics.StateSchema{
+			NumUint:      localSchemaUints,
+			NumByteSlice: localSchemaByteSlices,
+		}
+
+		globalSchema := basics.StateSchema{
+			NumUint:      globalSchemaUints,
+			NumByteSlice: globalSchemaByteSlices,
+		}
+
+		approvalProg, clearProg := mustParseProgArgs()
+
+		onCompletion := mustParseOnCompletion(createOnCompletion)
+
+		switch onCompletion {
+		case transactions.CloseOutOC, transactions.ClearStateOC:
+			reportWarnf("'--on-completion %s' may be ill-formed for 'goal app create'", createOnCompletion)
+		}
+
+		tx, err := client.MakeUnsignedApplicationCallTx(
+			appIdx, applicationArgs, appAccounts, foreignApps, foreignAssets,
+			onCompletion, approvalProg, clearProg, globalSchema, localSchema, 0)
+
+		if err != nil {
+			reportErrorf("Cannot create application txn: %v", err)
+		}
+
+		// Fill in note and lease
+		tx.Note = parseNoteField(cmd)
+		tx.Lease = parseLease(cmd)
+
+		// Fill in rounds, fee, etc.
+		fv, lv, err := client.ComputeValidityRounds(firstValid, lastValid, numValidRounds)
+		if err != nil {
+			reportErrorf("Cannot determine last valid round: %s", err)
+		}
+
+		tx, err = client.FillUnsignedTxTemplate(account, fv, lv, fee, tx)
+		if err != nil {
+			reportErrorf("Cannot construct transaction: %s", err)
+		}
+		explicitFee := cmd.Flags().Changed("fee")
+		if explicitFee {
+			tx.Fee = basics.MicroAlgos{Raw: fee}
+		}
+
+		// Broadcast
+		wh, pw := ensureWalletHandleMaybePassword(dataDir, walletName, true)
+		signedTxn, err := client.SignTransactionWithWallet(wh, pw, tx)
+		if err != nil {
+			reportErrorf(errorSigningTX, err)
+		}
+
+		txid, err := client.BroadcastTransaction(signedTxn)
+		if err != nil {
+			reportErrorf(errorBroadcastingTX, err)
+		}
+
+		// Report tx details to user
+		reportInfof("Issued transaction from account %s, txid %s (fee %d)", tx.Sender, txid, tx.Fee.Raw)
+
+		if !noWaitAfterSend {
+			_, err := waitForCommit(client, txid, lv)
+			if err != nil {
+				reportErrorf(err.Error())
+			}
+
+			resp, err := client.PendingTransactionInformationV2(txid)
+			if err != nil {
+				reportErrorf(err.Error())
+			}
+
+			if retTypeStr == "void" {
+				return
+			}
+
+			// specify the return hash prefix
+			hashRet := sha512.Sum512_256([]byte("return"))
+			hashRetPrefix := hashRet[:4]
+
+			var abiEncodedRet []byte
+			for i := len(*resp.Logs) - 1; i >= 0; i-- {
+				retLog := (*resp.Logs)[i]
+				if bytes.HasPrefix(retLog, hashRetPrefix) {
+					abiEncodedRet = retLog[4:]
+					break
+				}
+			}
+
+			if len(abiEncodedRet) == 0 {
+				reportErrorf("cannot find return log for abi type %s", retTypeStr)
+			}
+
+			retType, err := abi.TypeOf(retTypeStr)
+			if err != nil {
+				reportErrorf("cannot cast %s to abi type: %v", retTypeStr, err)
+			}
+			decoded, err := retType.Decode(abiEncodedRet)
+			if err != nil {
+				reportErrorf("cannot decode return value %v: %v", abiEncodedRet, err)
+			}
+			fmt.Println(decoded)
```

Comment:
Find the logged entry with `hash("return")` 4 bytes prefix, and parse with rules specified by ABI return type.